spec.proto

```diff
@@ -0,0 +1,55 @@
+/**
+ * CloudEvent Protobuf Format
+ *
+ * - Required context attributes are explicity represented.
+ * - Optional and Extension context attributes are carried in a map structure.
+ * - Data may be represented as binary, text, or protobuf messages.
+ */
+
+syntax = "proto3";
+
+package io.cloudevents.v1;
+
+import "google/protobuf/any.proto";
+import "google/protobuf/timestamp.proto";
+
+option java_package = "io.cloudevents.v1.proto";
+
+message CloudEvent {
+
+  // -- CloudEvent Context Attributes
+
+  // Required
+  string id = 1;
+  string source = 2; // URI-reference
+  string spec_version = 3;
+  string type = 4;
+
+  // Optional & Extension
```

Comment:
Why treating optional attributes similar to the extensions? How about we define optionals using proto optional fields (e.g. 'optional string subject = 5')?

Reply:
Agree with hrasadi@ in principle. 
I understand that with the latest proto spec there's no distinction between required vs optional fields and strictly speaking all fields are optional to set. However, I am not sure if that warrants abandoning the proto spec for fields that are 'optional to set' and instead put them inside one big map.

I think the way to model CloudEvents spec in proto would be that the fields that are 'optional to set' can have empty (default) values when a consumer reads that field. A writer can continue to not set an optional field if it doesn't need to (or when the value is empty).

// Required
string id = 1;  // must not be empty
...

// Optional
string subject = 5;  // empty, when not set
...

In such an implementation, there's no difference between when `subject` wasn't set and when `subject` was explicitly set an empty value (though writers shouldn't do that). In other words, empty value for `subject` means only one thing - an invalid value and not an actual subject. I believe, this will be in line with CloudEvents spec; at-least in principle.

Reply:
Per jskeet@ response on the main thread: 
Apparently, `optional` keyword is being reintroduced in proto3 for the sake of field presence check. So, I think it is even more compelling to have optional fields (like `subject`) directly in the proto as opposed to putting them in the `attributes` map.
https://github.com/protocolbuffers/protobuf/blob/master/docs/implementing_proto3_presence.md

Reply:
The concern here has really been around how the schema should evolve, it's possible that a future minor version of the spec might introduce a new 'optional' attribute. Formally declaring the 'required' CE attributes and then carrying the optional ones in the map structure seemed the least brittle way to allow evolution.

Reply:
Thanks for the reply. I guess my question is why not introduce it now? 

Totally unaware of other risks that you might have in mind, having the optional attributes as first class members of the proto improves the usability and reduces future toil of migration etc.

I think you mentioned there will be a conf. call. I am happy to provide more context there.

Reply:
Hi @JemDay, Thank you for moving this change forward!

Why is storing optional attributes in a map least brittle way? What's wrong with using the proto `optional` keyword? As mentioned in the last comment, it sounds like there's a clean and future-proof way of specifying optional attributes in the proto schema. So why not use the proto features as opposed to bypassing them and storing everything (optional) in a map.

My concern is that we are circumventing the proto spec and making it hard for every proto user that may want to consume CloudEvents. Rather than using proto generated methods, the user will have to access a loosely-typed map. Also, users will have to remember the right string values for the attribute names - which are not listed anywhere in this spec.

Reply:
The set of CloudEvents attributes is unbounded -- while the spec defines certain attributes, it also leaves open both public and private extensions which might be standardized later.

With respect to proto, there's no consistent requirement for registering tag values for these extensions (and since this would be a proto-unique requirement, I think @duglin has been reluctant to require this -- it also makes generic translation bridges between other formats and proto more difficult), so moving an extension from "allowed but not specifically known to proto" to "known to proto and assigned a tag number" is a **breaking** wire-format change.

Consider the case of promoting a "signature" attribute from extension to core. If we had a definition like so:

```
message CloudEvent {
  string datacontenttype = 5;
  // etc
  map<string, CloudEventAttribute> extensions = 12;
  // new standard attributes!
  bytes signature = 13;
}
```

Consider what recipients who need to handle messages from old and new senders will need to do:

```
// Oh look, we wrote our own accessor wrapper on top of Proto!
byte[] GetSignature(CloudEvent ce) {
  if ce.getSignature().Length() > 0 {
    return ce.getSignature();
  }
  return ce.getAttributesOrDefault("signature", new CloudEventAttribute()).getCloudEventBinary();
}
```

And the sender needs to encode the data in **both** locations if it doesn't know if all recipients have been upgraded yet:

```
// We also need to provide our own setters, which we might choose to do anyway to get type-safe attribute setters.
void setSignature(CloudEvent ce, byte[] signature) {
  ce.setSignature(signature)
  // Also set "signature" in extensions for clients who aren't aware of the signature field
  // Note that this doubles the storage size of these attributes until everything in the
  // ecosystem has upgraded. It also introduces a source-of-truth problem if the two
  // fields somehow ended up being set to different values.
  ce.putAttributes("signature", CloudEventAttribute.newBuilder().setCloudEventBinary(signature).Build());
}
```


@JemDay -- reading the above code, I think you may want to name `map<string, CloudEventAttribute>` as `attribute` rather than `attributes`, so that the code reads more pleasantly: `ce.putAttribute("signature", sig)` seems better than `ce.putAttributes("signature", sig)` for setting a single attribute.

Reply:
@evankanderson

Thanks, i was about to embark on crafting a similar explanation and you saved me some work :-)

I'll tweak the naming, the resultant code does read better - the proto map name might not!)

Reply:
@evankanderson 
Thanks for the details, this is very helpful! However, I disagree with the conclusion.

First, note that I was only asking to take the standard optional attributes out of the map not extension attributes that can be different for different providers. Leaving extensions attributes in the map LGTM.

My argument is twofold:
1. I challenge the premise that replacing an extension attribute with a standard attribute must be backwards compatible. I think that is impossible in real world because there's no semantic-consistency guarantee for an extension attribute named "foo" across providers.  
What is the rationale for this goal? And how will CNCF spec guarantee that an extension attribute "foo" has the same type and semantics across different event providers before graduating it to a standard optional attribute?
2. Even if the aforementioned is deemed as a worthy goal, merging the namespace of standard optional attributes and extension attributes by putting everything in a single map is not a good solution IMHO. This is because they have different scopes (extensions can be experimental or highly provider-specific). 

If an existing extension attribute "foo" is graduated to a standard attribute with the same name, how is the CNCF spec going to guarantee that the type and semantics of this new standard attribute is exactly the same as that of *all extension attributes named "foo"*? Otherwise, it'll force the providers to update their extension attribute (meaning a backwards incompatible change) - or opening the door for changes that can silently break existing user (meaning a silent change in semantics, format, etc). 

Overall, I still think that by bypassing proto spec we will end up accidentally punishing the proto users.